Use deep equality when comparing IngestDocument equality

dakrone opened the pull request:
This commit changes the `IngestDocument` class equality check to use "deep" equality. Previously we would end up doing the equivalent of `Objects.equals(thisMap, thatMap)` for the document. If a script or processor ended up writing any primitive arrays however, then those would fail the equality check because instance equality was used instead of check the actual values.

dakrone committed:
Use deep equality when comparing IngestDocument equality

This commit changes the `IngestDocument` class equality check to use "deep" equality. Previously we would end up doing the equivalent of `Objects.equals(thisMap, thatMap)` for the document. If a script or processor ended up writing any primitive arrays however, then those would fail the equality check because instance equality was used instead of check the actual values.

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

joegallo commented on server/src/main/java/org/elasticsearch/ingest/IngestDocument.java:
```diff
@@ -908,7 +908,7 @@ public boolean equals(Object obj) {
 
         IngestDocument other = (IngestDocument) obj;
-        return Objects.equals(ctxMap, other.ctxMap) && Objects.equals(ingestMetadata, other.ingestMetadata);
+        return Objects.equals(ctxMap, other.ctxMap) && Maps.deepEquals(ingestMetadata, other.ingestMetadata);
```
I don't think `Maps.deepEquals(...)` happens to work recursively -- if you have string keys and long[] values, it'll do a good job, but if you have string keys and map values (where the maps are string keys to long[] values) then it doesn't work.

I think the consequence for us here is that nested field in an ingest document would still throw this off (e.g. `"somearray"` versus `"nested.somearray"`).

dakrone replied:
Yes I agree. It doesn't work recursively. I will revisit this when I can.

joegallo replied:
https://github.com/elastic/elasticsearch/pull/98901 has been merged which fixes `Maps.deepEquals` to handle this case correctly, and I've added a commit here to make sure we're exercising the nested maps codepath (which would have failed prior to #98901).

joegallo reviewed (changes requested):
Github requires that I type words into this box (please see previously added comment-at-a-line).

joegallo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into ingest-document-deepequals

joegallo committed:
Use a nested array for this test

joegallo committed:
We control the schema for the metadata map

and I'm not aware of any circumstance where we'd have arrays in one,
so I think deepEquals comparison isn't necessary.

joegallo committed:
Merge branch 'main' into ingest-document-deepequals
